src/RA_StringUtils.h

```diff
@@ -24,9 +26,118 @@ _NODISCARD std::string Narrow(_In_ const std::string& wstr);
+                nCheck = std::vsnprintf(sFormatted.data(), sFormatted.capacity(), sFormat, pArgs);
+            else if constexpr (std::is_same_v<CharT, wchar_t>)
+                nCheck = std::vswprintf(sFormatted.data(), sFormatted.capacity(), sFormat, pArgs);
+            va_end(pArgs);
+            post_check(nCheck);
+        }
+
+        sFormatted.resize(ra::to_unsigned(nNeeded));
+    }
+
+    return sFormatted;
+}
+// This one does depend on the character set
+template<typename Arithmetic, class = std::enable_if_t<std::is_arithmetic_v<Arithmetic>>>
+_NODISCARD inline auto to_tstring(_In_ Arithmetic a) noexcept
+{
+#if _MBCS
+    return std::to_string(a);
+#elif _UNICODE
+    return std::to_wstring(a);
+#else
+#error Unknown character set detected, only MultiByte and Unicode character sets are supported!
+#endif // _MBCS
+} // end function to_tstring
+
+// More functions to be Unicode compatible w/o sacrificing MBCS
+/* A wrapper for converting a string to an unsigned long depending on _String */
+/* The template parameter does not need to be specified */
+template<typename CharT, typename = std::enable_if_t<ra::is_char_v<CharT>>>
+_Success_(0 < return < ULONG_MAX) _NODISCARD inline auto __cdecl
+tcstoul(_In_z_                   const CharT* __restrict str,
+        _Out_opt_ _Deref_post_z_ CharT**      __restrict endptr = nullptr,
+        _In_                     int                     base   = 10)
+{
+    const auto post_check = [](unsigned long ret) noexcept
+    {
+        assert((errno != ERANGE) && (ret != ULONG_MAX)); // out of range error
+        assert(ret != 0); // invalid argument, no conversion done
+    };
+    if constexpr (std::is_same_v<CharT, char>)
+    {
+        const auto ret = std::strtoul(str, endptr, base);
+        post_check(ret);
+        return ret;
+    }
+    if constexpr (std::is_same_v<CharT, wchar_t>)
+    {
+        const auto ret = std::wcstoul(str, endptr, base);
+        post_check(ret);
+        return ret;
+    }
+} /* end function tcstoul */
+
+/// <summary>
+///   Returns the number of characters of a null-terminated byte string or wide-character string depending on
+///   the character set specified
 /// </summary>
-_NODISCARD std::string StringPrintf(_In_z_ _Printf_format_string_ const char* const sFormat, ...);
+/// <param name="_Str">The string.</param>
+/// <returns>Number of characters</returns>
+/// <remarks>The template argument does not have to be specified.</remarks>
+template<typename CharT, typename = std::enable_if_t<ra::is_char_v<CharT>>>
+_NODISCARD _Success_(0 < return < strsz) inline auto __cdecl
+tcslen_s(_In_reads_or_z_(strsz) const CharT* const str, _In_ std::size_t strsz = UINT16_MAX) noexcept
+{
+    assert(str != nullptr);
+    const auto post_check = [&](std::size_t retVal) noexcept
+    {
+        assert(retVal != 0); // somehow str was a nullptr
+        assert(retVal != strsz); // str isn't null-terminated
+    };
```

Comment:
similar comment here. this method could be:
```
assert(str != nullptr);
int ret;
if constexpr (std::is_same_v<CharT, char>) {
    ret = strnlen_s(str, strsz);
} else if constexpr (std::is_same_v<CharT, wchar_t>) {
    ret = wcsnlen_s(str, strsz);
} else {
    #error Unsupported CharT
}
assert(ret != 0);
assert(ret != strsz);
return ret;
```
or, you could just have two methods and avoid the templating altogether. it's not gaining you anything in this particular case.
```
_NODISCARD int __cdecl tcslen_s(const char* str, std::size_t strsz = UINT16_MAX) noexcept
{
    int ret = strnlen_s(str, strsz);
    assert(ret != 0);
    assert(ret != strsz);
    return ret;
}
_NODISCARD int __cdecl tcslen_s(const wchar_t* str, std::size_t strsz = UINT16_MAX) noexcept
{
    int ret = wscnlen_s(str, strsz);
    assert(ret != 0);
    assert(ret != strsz);
    return ret;
}
```

Reply:
I already have std::enable_if_t, the only character types that are allowed are `char `and `wchar_t`. If that condition failed, the dll won't even compile because it's a compile-time test.
They have something similar in C# but with different syntax.
`where T : struct`

Would it be easier for you to follow if I did `static_assert `instead of std::enable_if_t.

detail version:
```
template<typename CharT>
struct _NODISCARD is_char :
    std::bool_constant<(std::is_same_v<CharT, char> || std::is_same_v<CharT, wchar_t>)>
{
};
```
User version:
```
template<typename CharacterType> _CONSTANT_VAR
is_char_v{ detail::is_char<CharacterType>::value };
```